### ISSUE TITLE
Cannot bind to UNIX socket

### ISSUE THREAD

**Issue by siliconsheep** [cannatag/ldap3#803]
Hi!

I cannot seem to bind to a local OpenLDAP server using a UNIX socket. Even using the sample code from the documentation gives me an error:

Python 3.6.8 (default, Aug  7 2019, 17:28:10) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ldap3 import Server, Connection, SASL, EXTERNAL
>>> s = Server('ldapi://var/run/ldapi')
>>> c = Connection(s, authentication=SASL, sasl_mechanism=EXTERNAL, sasl_credentials='')
>>> c.bind()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/ldap3/core/connection.py", line 561, in bind
    self.open(read_server_info=False)
  File "/usr/local/lib/python3.6/site-packages/ldap3/strategy/sync.py", line 56, in open
    BaseStrategy.open(self, reset_usage, read_server_info)
  File "/usr/local/lib/python3.6/site-packages/ldap3/strategy/base.py", line 139, in e
    raise exception_history[0][0]
ldap3.core.exceptions.LDAPSocketOpenError: a bytes-like object is required, not 'tuple'

Any idea what's going on?

Thanks!
Dieter

**Comment by kpinc** [comment]
Hi,

The problem is at line 216 of the _open_socket() function in strategy/base.py.

It is not just unix sockets that are broken.  From reading the code it
also looks like IPv6 is also broken.

Broken in commit: 35a82870973

The bind call is with a pair (a tuple) and when you've a Unix socket
you get the following exception:

  TypeError: a bytes-like object is required, not 'tuple'

A Unix socket can't be called with a tuple, it needs the path to the
socket.

Something like this instead of the current line 216 might fix it:

```
if address[0] == socket.AF_UNIX:
    self.connection.socket.bind(address[4])
else:
    self.connection.socket.bind((self.connection.source_address, source_port))
```

But I've not the time to do _any_ testing and work up a patch, nor is
my proposed code particularly self-explanatory.  And testing just for
AF_UNIX is probably also no good, since AF_INET6 (at least) should be
supported.  Nor have I looked into how 2.6.1 is really working and
whether the code should instead fall-through to the old behavior for
unix sockets/IPv6/whatever else.


This is not a problem with ldap3 version 2.6.1, but is a problem with 2.7.

The work-around is to not use unix sockets but instead use localhost.

Regards,
Karl

**Comment by cannatag** [comment]
Hi, can you try the code in the dev branch? LDAPI should work again.

Thanks,
Giovanni

**Comment by cannatag** [comment]
Fixed in ldap3 v2.8.
